WooCommerce/Classes/Tools/InfiniteScroll/PaginationTracker.swift

```diff
@@ -0,0 +1,152 @@
+import Yosemite
+
+/// Delegate of `PaginationTracker` that implements syncing per page number and size.
+protocol PaginationTrackerDelegate: class {
+    typealias SyncCompletion = (Result<Bool, Error>) -> Void
+
+    /// Syncs a page by the number and size, and returns an error or whether there might be the next page on completion.
+    ///
+    /// - Parameter reason: A value passed from `resync` or `syncFirstPage`. This can
+    ///                     be used to decide how to perform the sync.
+    /// - Parameter onCompletion: called on sync completion that returns either an error or a boolean that indicates
+    ///                           whether there might be the next page to sync.
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?)
+}
+
+/// Keeps track of the pagination for API syncing to support infinite scroll.
+final class PaginationTracker {
+    /// Default Settings
+    private enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
+        static let pageSize = 25
+    }
+
+    /// First Page Index
+    private let pageFirstIndex: Int
+
+    /// Number of elements retrieved per request.
+    private let pageSize: Int
+
+    /// Syncing delegate
+    weak var delegate: PaginationTrackerDelegate?
+
+    /// Maps from a page number to the last synced date.
+    private var syncedDatesByPage = [Int: Date]()
+
+    /// Indexes of the pages being currently synced
+    private var pagesBeingSynced = IndexSet()
+
+    /// Whether there might be more pages to fetch from the API, set by the sync completion.
+    private var hasNextPage: Bool = true
+
+    /// Returns the highest page number that is currently being synced, if any.
+    private var highestPageBeingSynced: Int? {
+        pagesBeingSynced.max()
+    }
+
+    /// Returns the highest page number that has been successfully synced.
+    private var highestPageSynced: Int? {
+        syncedDatesByPage.keys.max()
+    }
```

Comment:
What do you think about making this return zero by default instead? 

```suggestion
    private var highestPageSynced: Int {
        syncedDatesByPage.keys.max() ?? pageFirstIndex - 1
    }
```

Reply:
I felt a bit odd to return a non-nil value when no pages have been synced yet 🤔 does it look okay to you to keep it optional?

Reply:
Yeah, I guess the optional has some value too. I think I just commented on this because it took me a while to understand. 😅 But all good!